src/sysupdate/sysupdate-transfer.c

```diff
@@ -779,43 +782,196 @@ static void compile_pattern_fields(
         memcpy(ret->sha256sum, i->metadata.sha256sum, sizeof(ret->sha256sum));
 }
 
+static int helper_on_exit(sd_event_source *s, const siginfo_t *si, void *userdata) {
+        const char *name = userdata;
+        int code;
+
+        assert(s);
+        assert(si);
+        assert(name);
+
+        if (si->si_code == CLD_EXITED) {
+                code = si->si_status;
+                if (code != EXIT_SUCCESS)
+                        log_error("%s failed with exit status %i.", name, code);
+                else
+                        log_debug("%s succeeded.", name);
+        } else {
+                code = -EPROTO;
+                if (IN_SET(si->si_code, CLD_KILLED, CLD_DUMPED))
+                        log_error("%s terminated by signal %s.", name, signal_to_string(si->si_status));
+                else
+                        log_error("%s failed due to unknown reason.", name);
+        }
+
+        return sd_event_exit(sd_event_source_get_event(s), code);
+}
+
+struct notify_userdata {
+        const Transfer *transfer;
+        const Instance *instance;
+        const TransferProgress callback;
+        pid_t pid;
+        void* userdata;
+};
+
+static int helper_on_notify(sd_event_source *s, int fd, uint32_t revents, void *userdata) {
+        char buf[NOTIFY_BUFFER_MAX+1];
+        struct iovec iovec = {
+                .iov_base = buf,
+                .iov_len = sizeof(buf)-1,
+        };
+        CMSG_BUFFER_TYPE(CMSG_SPACE(sizeof(struct ucred)) +
+                         CMSG_SPACE(sizeof(int) * NOTIFY_FD_MAX)) control;
```

Comment:
you don't intend to receive fds here, hence drop the second CMSG_SPACE() expression